### PR TITLE
add keepat! for in-place logical filtering

### LINKED ISSUE #39470: Inplace mask of vector
It was recently asked on Slack if there is an inplace equivalent of `my_vector = my_vector[my_mask]`, where we assume that the user already needs the mask and does not want to essentially recompute it with `filter!`. I took a glance at the implementation of `filter!` in Base and saw that it could be easily adapted to this use-case:
```julia
function mask!(a::AbstractVector, m::AbstractVector{Bool})
    j = firstindex(a)
    for i in eachindex(a, m)
        @inbounds begin
            ai = a[i]
            mi = m[i]
            a[j] = ai
        end
        j = ifelse(mi, nextind(a, j), j)
    end
    j > lastindex(a) && return a
    if a isa Vector
        resize!(a, j-1)
        sizehint!(a, j-1)
    else
        deleteat!(a, j:lastindex(a))
    end
    return a
end

julia> let a = [:a, :b, :c], m = [true, false, true]
           mask!(a, m)
       end
2-element Vector{Symbol}:
 :a
 :c
```

Is this something that would be worth putting in Base? If so, I can whip up a PR, but don't want to waste my time if it's something that's already been discussed. I tried searching around but didn't find anything, but perhaps I missed a discussion.

### PR DESCRIPTION
**Pull request by MasonProtter**
Closes #39470

### CODE REVIEW

**Comment by kshyatt** [comment]
Could doc strings be added for this? Otherwise lgtm

**Review by kshyatt — DISMISSED** [review]
With docs this should be good to go I think

**Comment by MasonProtter** [comment]
Done! I also added a `NEWS.md` entry. Let me know if the descriptions need tweaking.

File: base/array.jl [hunk 1]
```diff
@@ -2436,6 +2436,48 @@ function filter!(f, a::AbstractVector)
...
+"""
+function mask!(a::AbstractVector, m::AbstractVector{Bool})
+    j = firstindex(a)
+    for i in eachindex(a, m)
+        @inbounds begin
+            ai = a[i]
+            mi = m[i]
+            a[j] = ai
+        end
+        j = ifelse(mi, nextind(a, j), j)
```
Comment (vtjnash) [inline base/array.jl:2469]: nit: there might be non-zero cost to copying an element, which the `ifelse` function implies (and I don't believe there's any simd optimization possible here due to the loop carried dependencies)
```suggestion
            if m[i]
                i == j || (a[j] = a[i])
                j = nextind(a, j)
            end
        end
```

File: base/array.jl [hunk 2]
```diff
@@ -2436,6 +2436,48 @@ function filter!(f, a::AbstractVector)
...
+    end
+    j > lastindex(a) && return a
+    if a isa Vector
+        resize!(a, j-1)
+        sizehint!(a, j-1)
+    else
+        deleteat!(a, j:lastindex(a))
+    end
```
Comment (vtjnash) [inline base/array.jl:2477]: resize! and deleteat! call the same underlying method, so there's not really any reason to split the code path
```suggestion
    deleteat!(a, j:lastindex(a))
```

**Review by vtjnash — APPROVED** [review]
Can you add a reference to this into doc/src/base/arrays.md

File: base/array.jl (hunk 1 above)
Comment (MasonProtter) [reply base/array.jl:2469]: Interesting, should `filter!` also get an analogous change then? https://github.com/JuliaLang/julia/blob/master/base/array.jl#L2441

File: base/array.jl (hunk 2 above)
Comment (MasonProtter) [reply base/array.jl:2477]: This is also something I lifted from the implementation of `filter!`. Should this be changed in `filter!` as well?

**Comment by MasonProtter** [comment]
I can add this to `doc/src/base/arrays.md`, but maybe it'd fit better in doc/src/base/collections.md as that is where `filter!` is?

File: base/array.jl (hunk 1 above)
Comment (vtjnash) [reply base/array.jl]: Ah, interesting. yeah, seems like that could be a good change. This has already crossed to https://github.com/JuliaLang/julia/pull/36229 also. And might have some implications for https://github.com/JuliaArrays/OffsetArrays.jl/pull/137 already too?

**Comment by vtjnash** [comment]
yeah, that sounds good

File: base/array.jl [hunk 3]
```diff
@@ -2450,6 +2450,43 @@ function filter!(f, a::AbstractVector)
...
+
+julia> a
+2-element Vector{Symbol}:
+ :a
+ :c
+```
+"""
+function mask!(a::AbstractVector, m::AbstractVector{Bool})
+    j = firstindex(a)
+    for i in eachindex(a, m)
```
Comment (vtjnash) [inline base/array.jl:2477]: Q for anyone: is `eachindex` and `nextind` / `iterate` guaranteed to be in the same visit order?

Comment (vtjnash) [reply base/array.jl:2477]: @mbauman @timholy @JeffBezanson?

**Comment by rfourquet** [comment]
Shouldn't we unify the names with `keepat!` from #36229? For consistency with `deleteat!` which accepts either a mask or indices. I think `mask!` wouldn't work as a name for the `keepat!` functionality, would `keepat!` work as a name for the `mask!` functionality?

**Comment by musm** [comment]
> Shouldn't we unify the names with `keepat!` from #36229? For consistency with `deleteat!` which accepts either a mask or indices. I think `mask!` wouldn't work as a name for the `keepat!` functionality, would `keepat!` work as a name for the `mask!` functionality?

Good point to bring this up. Are these two PRs stalled on bikeshedding a name?

**Comment by simeonschaub** [comment]
Removing merge me label and marking for triage since it looks like there still has to be a definite decision made on the name.

**Comment by oxinabox** [comment]
Is this just because `deleteat(xs, .!mask)` is to expensive because flipping the mask?

**Comment by oscardssmith** [comment]
triage says `keepat!`

**Comment by MasonProtter** [comment]
Okay, I'll amend the PR with the new name, sorry I couldn't make the Triage call.

**Comment by rfourquet** [comment]
@MasonProtter did you close this by inadvertently?

**Comment by StefanKarpinski** [comment]
Reopened in case. Can close again if it was on purpose.

**Comment by MasonProtter** [comment]
I closed this because https://github.com/JuliaLang/julia/pull/36229 was merged, which is the exact same thing, right? I didn’t realize they were doing the same thing until yours was merged @rfourquet or I wouldn’t have opened this one in the first place.

**Comment by vtjnash** [comment]
I believe they do different, but complimentary things

**Comment by rfourquet** [comment]
Yes IIRC, the method from the other PR was receiving indices, while this one here receives a boolean mask.

**Comment by MasonProtter** [comment]
🤦 well im glad you paid closer attention than I did!

**Comment by oscardssmith** [comment]
Triage decided on different methods a few weeks ago when we discussed this. The reason is that deleteat has methods for both.

**Comment by bkamins** [comment]
will this get 1.7 label? I would like to add `keepat!` to https://github.com/JuliaLang/Compat.jl/issues/750 and it would be great to know what contract it will support in 1.7. Thank you!

**Comment by MasonProtter** [comment]
Superseded by #42351

**Pull request closed without merging**